Update the OTA PAL tests to use OTA v3.0.0

yanjos-dev opened the pull request:
Update OTA PAL tests to use OTA v3.0.0

Note
----------
The build checks are failing for this PR, even for projects that are unrelated to OTA. This was not introduced in this PR. This is because of the new ota_appversion32.h header file that was introduced in V3.0.0 of the OTA library. This causes non-OTA demos to break because it is related to the common demo header files. Fixing this will require updating non-OTA files as well as nearly all of the projects. This will be fixed in a following PR to limit the size and content of this PR.

Description
-----------
The OTA library was updated from v2.0.0 to v3.0.0. This major release changed the functionality and API of the OTA PAL. This PR is to update the qualification tests for the OTA PAL that are ran to qualify ports.

The original test files were removed before making this change, so I added them back in before modifying them. Because they were originally removed, it will be hard to tell what changed in this PR. I recommend doing a diff with the first commit on this branch ( 4d6586e9dbc48c4779697626bccf194c080dc7ed ) where I add back in the original files with the latest commit on this branch.

Overview of changes
-----------
These are the **major changes made**:
* Removed the "Full_OTA_AGENT" test group along with all related files. This test group was previously required for qualification. These tests were for the OTA library and not specific to the OTA PAL port. The OTA library now has full test coverage that is hosted in the OTA repository so this test group is no longer required.
* Removed the "Full_OTA_CBOR" and "Quarantine_OTA_CBOR" test groups as well as all related files. These tests were not part of the qualification tests. The functionalities these tests covered are now being tested in the OTA repository.
* Moved the testing files from the library directory to the tests/integration_tests/ota_pal directory.
* Updated the OTA PAL qualification tests to use v3.0.0 of the OTA library.
* Updated how the OTA PAL tests access the code signing certificate for tests. Previously there was a dedicated header file for the code signing credential. This is no longer the case for the new version of the library. I've updated the test code to instead expect this variable to be defined in ota_pal.c. The value is assigned to a macro that is defined in the platform specific OTA config file.
* Updated the test related files for the OTA PAL for ESP32 and WinSim platforms. Both are passing all tests now.

**Note:** There were no functional changes to the actual tests. They were only ported to the new API.

**Minor Changes:**
* Updated the testing README.md.
* Updated some print statements to use the logging stack.
* Renamed the test config file name.
* Updated CMake based on the latest file locations.
* Updated the Visual Studio test project based on latest changes.
* Updated the ESP32 OTA PAL port source code and build files based on the latest changes.

**Note:** This PR should be squashed before merging.

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have tested my changes. No regression in existing tests.
- [x] My code is Linted.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

yanjos-dev committed:
Breaking commit - Copy original test files back

This commit copies the OTA testing files from the 202012.00 release.
These files were originally in the libraries/freertos_plus/ota
directory. I'm moving these files back in so that it is clear what
changes are being made to the original test files. This will not compile
as is.

yanjos-dev committed:
Update file type and PAL functions for OTA tests

* Update the file context type for v3.0.0 of the OTA library.
* Update the name of the PAL functions and the parameters for v3.0.0 of
  the OTA library.

yanjos-dev committed:
Update error codes in the OTA PAL tests

* Error codes were changed in v3.0.0 of the OTA library. Update the
  OTA PAL qualification tests to reflect this.

yanjos-dev committed:
Update OTA PAL test includes for OTA v3.0.0

yanjos-dev committed:
Update OTA PAL test print statements for OTA v3

* Update the print statements to use the logging stack instead of the
  configPRINTF macro.

yanjos-dev committed:
Update OTA PAL test usage of structs for OTA v3

yanjos-dev committed:
Update additional OTA PAL test errors for OTA v3

yanjos-dev committed:
Move OTA PAL test helper functions

yanjos-dev committed:
Add header and fix image enum for OTA PAL tests

yanjos-dev committed:
Rename folder for OTA PAL tests

yanjos-dev committed:
Fix CMake build for OTA demo on winsim

yanjos-dev committed:
Update the name for OTA config file

yanjos-dev committed:
Remove deprecated header file from OTA PAL

yanjos-dev committed:
Update OTA PAL tests to be able to build for pc

* Update the name of the OTA PAL functions in the tests.
* Add to the build files and add a header file to enable access to the
  OTA PAL static functions in the OTA PAL port for PC.
* Add a ota_pal.h that will be used in place of the ports ota_pal.h.
  This is to enable setting the signature.

yanjos-dev committed:
Remove deprecated OTA PAL test files

yanjos-dev committed:
Add header files to OTA PAL test cmake

yanjos-dev committed:
Update OTA test project for PC

* Update files listed in the test project for v3.0.0 of the OTA library.
* Remove a post build command to copy over a folder that does not exist
  anymore.
* Add a missing firmware version variable to the PAL test file.

yanjos-dev committed:
Fix error handling in aws_test_ota_pal.c

yanjos-dev committed:
Add example creds back to the OTA test dir

yanjos-dev committed:
Update comments in OTA PAL test files

yanjos-dev committed:
Remove unused files from OTA PAL tests

yanjos-dev committed:
Remove unused macros from the test runnner config

yanjos-dev committed:
Cleanup the PC OTA PAL test config file

* Remove two unused macros.
* Update the comment content and formatting.
* Update the default path to the code signing certificate.

yanjos-dev committed:
Add header file to OTA PAL tests for vTaskDelay

yanjos-dev committed:
Update ota_pal_test.h for OTA v3

yanjos-dev committed:
Update OTA PAL test readme for OTA v3

yanjos-dev committed:
Sync the WinSim test project with CMake

yanjos-dev committed:
Uncrustify OTA PAL test files

yanjos-dev committed:
Get ESP32 qualification tests to all pass

This commit includes various changes that were required to get the ESP32
OTA PAL code to pass the qualification tests.
* Update return value of test_otaPal_CheckFileSignature to be
  OtaPalStatus_t.
* Update the OTA PAL tests to expect otaPal_CheckFileSignature to return
  OtaPalStatus_t instead of OtaPalMainStatus_t.
* Update the ESP32 CMakeLists.txt file to include the integration_test
  directory. This is required to include the "test_otaPal_CheckFileSignature"
  and "test_otaPal_ReadAndAssumeCertificate" functions from the qualification
  tests.
* Include the "aws_ota_pal_test_access_define.h" header file to the ota_pal.c
  file for ESP32.
* Remove the deprecated otatestVALID_SIG_METHOD and otatestINVALID_SIG_METHOD
  macros from aw_test_ota_config.h.
* Update "prvPAL_ReadAndAssumeCertificate" to "otaPal_ReadAndAssumeCertificate".
* Update "prvPAL_CheckFileSignature" to "otaPal_CheckFileSignature".
* Fix incorrect error values in ota_pal.c for esp32. Some of these errors had
  previously been accidentally set to OtaPalRxFileCreateFailed.
* Rename the OTA library config file to be "ota_config.h".

yanjos-dev committed:
Update the OTA PAL certificate management

* Move the code signing certificate variable to ota_pal.c. Have it set
  a macro that is set in the the platform specific config file.
* Delete the ota_pal_test.h file. This is no longer needed due to moving
  the code signing certificate variable.
* Delete the aws_test_ota_signature_methods.h file. This is no longer
  needed.

* Note: This commit does not change this for all platforms. Only the
  common to all platforms and the ESP32 platform have been updated.

yanjos-dev committed:
Comment testing include in ESP32 OTA PAL port

yanjos-dev committed:
Update OTA PAL cert. variable location for WinSim

* Move the certificate variable from the ota_pal.h file to the ota_pal.c
  file.

yanjos-dev committed:
Remove deprecated files from the WinSim test proj.

yanjos-dev committed:
Uncrustify OTA PAL test files

yanjos-dev committed:
Uncrustify ESP32 OTA PAL port files

yanjos-dev committed:
Uncrustify WinSim OTA PAL port files

yanjos-dev committed:
Update the OTA PAL test readme

yanjos-dev committed:
Update the ota_pal.h comments and ifdef

yanjos-dev committed:
Merge upstream into yanjos/pal_v3_qual_tests

yanjos-dev committed:
Add new config files to ESP32 and WinSim platforms

* Add a config file for the OTA demo.
* Add missing config files for the core_mqtt_agent_config.h header file.

yanjos-dev committed:
Move code signing cert. to demo/test config files

The OTA PAL uses a macro to define a code signing certificate. This
change moves the macro to a location common to the demo and test files.
The macro is now being defined in the ota_config.h file via either a
testing config file or a demo config file that is being included.

This change included the following sub-changes:
* Update the OTA library CMake target to depend on the OTA MCU port.
* Update the macro name of the code signing certificate.
* Update the ESP32 OTA MCU port dependencies.
* Add the ota_demo_config.h include to the ota_config.h for both esp32
and winsim demos.
* Update the esp32 and winsim ota_config.h files that are part of the
test projects to include the test config file.
* Update the ota_pal.c ports for esp32 and winsim to include
ota_config.h.
* Clean up artifacts of a previous git merge.

yanjos-dev committed:
Update the OTA submodule pointer to be on main

* This update is specifically to get a bug fix for the OTA PAL error
macro.

yanjos-dev committed:
Update the OTA PAL test README.md file

yanjos-dev committed:
Update the PC projects for the OTA demos and tests

yanjos-dev committed:
Merge upstream into yanjos/pal_v3_qual_tests

pvyawaha commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.c:
```diff
@@ -53,43 +61,42 @@ static inline BaseType_t prvContextValidate( OtaFileContext_t* pFileContext )
-				if( 0 == lFileCloseResult )
-				{
-					LogInfo( ( "File closed.\r\n" ) );
+                lFileCloseResult = fclose( C->pFile ); /*lint !e482 !e586
```
Is it required to close ? If the file is open can we use the same handle ?

yanjos-dev replied:
This was not a change I made so I can't answer this for sure without looking into it more. There is only a diff here because I ran uncrustify.

Lets address this in a separate PR since this would be a functional change unrelated to this PR.

yanjos-dev replied:
We agreed offline that this is not an issue so I'm resolving it.

yanjos-dev commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.h:
```diff
@@ -24,38 +24,34 @@
  *
- * Aborts access to an existing open file represented by the OTA file context C. This is only valid
- * for jobs that started successfully.
+ * Aborts access to an existing open file represented by the OTA file context pFileContext. This is
```
I believe I copied this from a separate location originally. I agree this should be consistent. Do we want C to be the name for this parameter in the definition though? I'm not sure I understand where the letter C comes from and what it represents. Would you be okay with me changing this to pFileContext everywhere?

yanjos-dev replied:
Fixed.

yanjos-dev commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.c:
```diff
@@ -34,16 +34,24 @@
+/* Include config file used for testing the OTA PAL. This config file defines
+ * the otapalconfigCODE_SIGNING_CERTIFICATE macro. */
+#ifdef FREERTOS_ENABLE_UNIT_TESTS
+    #include "aws_test_ota_config.h"
```
Good catch, this was left in on accident. It should just be:
```
#include "ota_config.h"
```

yanjos-dev replied:
Fixed.

yanjos-dev commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.c:
```diff
@@ -34,16 +34,24 @@
 const char OTA_JsonFileSignatureKey[ OTA_FILE_SIG_KEY_STR_MAX_LENGTH ] = "sig-sha256-ecdsa";
 
-static OtaErr_t otaPal_CheckFileSignature( OtaFileContext_t * const C );
+static const char signingcredentialSIGNING_CERTIFICATE_PEM[] = otapalconfigCODE_SIGNING_CERTIFICATE;
```
This was how it was before so I left it. I'm okay to change this though. Do you have any recommendations? I'm not sure what other device credentials you are referring to.

yanjos-dev replied:
Will change this to "codeSigningCertificatePEM"

yanjos-dev replied:
Done.

yanjos-dev committed:
Update the PC OTA PAL file context variable name.

* Change the variable name from C to pFileContext

yanjos-dev committed:
Include ota_pal.h in the OTA PAL test code

yanjos-dev committed:
Update PC PAL code signing macro include

* Update the OTA PAL port for Windows to include the
otapalconfigCODE_SIGNING_CERTIFICATE via ota_config.h.

yanjos-dev committed:
Update PC OTA PAL comments and signatures

yanjos-dev committed:
Update documentation in ESP32 ota_pal.h

pvyawaha commented on vendors/espressif/boards/ports/ota_pal_for_aws/ota_pal.h:
```diff
@@ -24,107 +24,106 @@
  *
- * Aborts access to an existing open file represented by the OTA file context C. This is only valid
- * for jobs that started successfully.
+ * Aborts access to an existing open file represented by the OTA file context pFileContext. This is
```
Please fix the doxygen here with arg names

yanjos-dev committed:
Update ESP32 OTA port file context variable name

pvyawaha commented on vendors/espressif/boards/esp32/aws_tests/config_files/core_mqtt_agent_config.h:
```diff
@@ -0,0 +1,71 @@
+ * @note Specified in bytes.  Must be large enough to hold the maximum
+ * anticipated MQTT payload.
+ */
+#define MQTT_AGENT_NETWORK_BUFFER_SIZE    ( 5000 )
```
This buffer will not be sufficient for ota over http

yanjos-dev replied:
This is okay to leave for the test config file. The OTA PAL tests do not use OTA over HTTP in any way. I believe this file was only required for building.

pvyawaha commented on tests/integration_test/ota_pal/test_files/rsa-sha256-signer.key.pem:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Why are we adding private keys here ? We should remove this and have instructions on how to generate.

pvyawaha commented on tests/integration_test/ota_pal/test_files/rsa-sha1-signer.key.pem:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
+MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQC5craIexo46oq3
```
Another private key. please see comment below.

pvyawaha commented on tests/integration_test/ota_pal/test_files/rsa-sha1-root-ca-cert-key.pem:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Private key?

pvyawaha commented on tests/integration_test/ota_pal/test_files/ecdsa-sha256-signer.key.pem:
```diff
@@ -0,0 +1,5 @@
+-----BEGIN PRIVATE KEY-----
```
Another one

yanjos-dev committed:
Update the OTA PAL code signing cert variable name

* Update the name from signingcredentialSIGNING_CERTIFICATE_PEM to
  codeSigningCertificatePEM in the OTA PAL tests, Windows and ESP32
  OTA PAL ports.

yanjos-dev committed:
Remove deprecated tests from the esp32 config

yanjos-dev commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.h:
```diff
@@ -64,68 +60,70 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C );
+ * @brief Authenticate and close the underlying receive file in the specified OTA context.
  *
- * @note The input OtaFileContext_t C is checked for NULL by the OTA agent before this
+ * @note The input OtaFileContext_t pFileContext is checked for NULL by the OTA agent before this
```
Fixed.

yanjos-dev commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.h:
```diff
@@ -64,68 +60,70 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C );
  * @brief Write a block of data to the specified file at the given offset.
  *
- * @note The input OtaFileContext_t C is checked for NULL by the OTA agent before this
+ * @note The input OtaFileContext_t pFileContext is checked for NULL by the OTA agent before this
```
Fixed.

yanjos-dev commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.h:
```diff
@@ -64,68 +60,70 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C );
  *
- * @return The number of bytes written on a success, or a negative error code from the platform abstraction layer.
+ * @return The number of bytes written on a success, or a negative error code from the platform
+ * abstraction layer.
```
Done.

yanjos-dev commented on vendors/pc/boards/windows/ports/ota_pal_for_aws/ota_pal.h:
```diff
@@ -64,68 +60,70 @@ OtaPalStatus_t otaPal_Abort( OtaFileContext_t * const C );
+ * partition or file must be completely erased or overwritten in this routine.
  *
- * @note The input OtaFileContext_t C is checked for NULL by the OTA agent before this
+ * @note The input OtaFileContext_t pFileContext is checked for NULL by the OTA agent before this
```
Done.

yanjos-dev committed:
Remove unnecessary OTA PAL test files
